src/WebJobs.Script/ExtensionBundle/ExtensionBundleManager.cs

```diff
@@ -201,30 +202,87 @@ private string GetBundleFlavorForCurrentEnvironment()
+                await (await response.Content.ReadAsStreamAsync()).CopyToAsync(stream);
+
+                _logger.DownloadComplete(zipUri, filePath);
+                return true;
+            }
+            catch (HttpRequestException ex)
             {
-                _logger.ErrorDownloadingZip(zipUri, response);
+                var statusCode = ex.StatusCode;
+                _logger.ErrorDownloadingZipDiskSize(
+                    ex,
+                    zipUri,
+                    statusCode,
+                    ex.HttpRequestError,
+                    filePath,
+                    GetDiskUsageSafe(filePath));
                 return false;
             }
-
-            using (var content = await response.Content.ReadAsStreamAsync())
-            using (var stream = new FileStream(filePath, FileMode.Create, FileAccess.Write, FileShare.None, bufferSize: 4096, useAsync: true))
+            catch (IOException ex)
+            {
+                _logger.ErrorDownloadingZipIO(
+                    ex,
+                    zipUri,
+                    filePath,
+                    GetDiskUsageSafe(filePath));
+                return false;
+            }
+            catch (Exception ex)
             {
-                await content.CopyToAsync(stream);
+                // Non-HttpRequestException path: log as unexpected without HTTP-specific fields.
+                _logger.ErrorDownloadingZipUnexpected(
+                    ex,
+                    zipUri,
+                    filePath,
+                    GetDiskUsageSafe(filePath));
+                return false;
             }
+        }
 
-            _logger.DownloadComplete(zipUri, filePath);
-            return true;
+        private string GetDiskUsageSafe(string path)
+        {
+            try
+            {
+                var root = Path.GetPathRoot(path);
+                if (string.IsNullOrEmpty(root))
+                {
+                    return "error=RootPathNotFound";
+                }
+
+                var di = new DriveInfo(root);
+                const double BytesPerMB = 1024d * 1024d;
+                double freeMb = di.AvailableFreeSpace / BytesPerMB;
+                double totalMb = di.TotalSize / BytesPerMB;
+                return $"free={freeMb:F2}MB total={totalMb:F2}MB";
+            }
+            catch (Exception ex)
+            {
+                return FormatDiskError(ex);
+            }
         }
 
-        private async Task<string> GetLatestMatchingBundleVersionAsync()
+        private static string FormatDiskError(Exception ex)
         {
-            using (var httpClient = new HttpClient())
+            var msg = ex.Message?.Replace(Environment.NewLine, " ").Trim();
+            if (!string.IsNullOrEmpty(msg) && msg.Length > 200)
             {
-                return await GetLatestMatchingBundleVersionAsync(httpClient);
+                msg = msg.Substring(0, 200) + "...";
```

Comment:
Any particular reason for 200 length?

Reply:
The formatting here is not ideal. Not a blocker, but those concerns (e.g., logging exception format) would be handled more generically.